### DISCUSSION TITLE
Usage as a library (maybe with a conan package)

### DISCUSSION THREAD

**Discussion by ruanformigoni** [mhx/dwarfs#189] [General]
Greetings,

Is it possible to use dwarfs as a library? It would be great if this project could be published to [conan](https://conan.io/), that way creating C or C++ project that use dwarfs would be quite trivial.

**Comment by mhx** [comment]
Hi @ruanformigoni!

DwarFS is designed as a library, but there's no guarantee on a stable interface before v1.0.0. In fact, at the moment, *a lot* of interface refactoring happens between releases. The [API to use DwarFS images](https://github.com/mhx/dwarfs/blob/main/include/dwarfs/filesystem_v2.h) is relatively stable, though.

I'm going to be honest and say that I don't know much about conan other than that it exists, so I can't judge how much effort it would be to publish/maintain it on that platform.

This is definitely something I'd consider as the API stabilizes. If you have a concrete use case and don't mind breaking API changes, I'm happy to accept PRs in case any changes to DwarFS are necessary.

**Reply by ruanformigoni** [reply]
Thanks for the reply @mhx , for my use case I use dwarfs in two of my projects, [flatimage](https://github.com/ruanformigoni/flatimage) and [gameimage](https://github.com/ruanformigoni/gameimage/). They are proof of concepts that I prototyped in bash, and now I'm porting to C++.  

Here is a snippet of how I'm using it at the moment with boost::process:
```cpp
  // Compress
  ns_subprocess::subprocess(str_image
    , "fim-exec"
    , "mkdwarfs"
    , "-f"
    , "-i"
    , str_app
    , "-o"
    , str_target
  );
```

It works, but I have to include the dwarfs binary separately, and I want my application to be a single binary. I'll try to integrate with your CMakeLists.txt and provide feedback on what I can. A pre-compiled libdwarfs.a in the releases would help a lot as well, if you'd like I could work on that.

Lastly, would you consider compiling with `musl libc` instead of `glibc`? The `glibc` binary works, but it has a licensing issue since it is provided as LGPL, so the object code must be provided and static linking voids that ([here](https://opensource.stackexchange.com/questions/13588/how-should-i-license-my-software-if-it-links-statically-with-glibc) is a summary). I have implemented some time ago a [musl dockerfile](https://github.com/ruanformigoni/dwarfs/blob/main/docker/Dockerfile.alpine) for dwarfs. My memory is a bit faint on that right now, but I could help to change the project libc to `musl libc` if you'd like.

**Reply by mhx** [reply]
Hi, sorry for dropping the ball on this.

I'd be happy to integrate pull requests for musl compatibility, so feel free to give it a shot.

Regarding a dwarfs library, I don't currently have plans to provide one. That's not to say I'm not going to change my mind, but I need a bit more time to think about what I want the interfaces to look like. As mentioned before, the API to *access* images would be relatively straightforward, but the API to *build* images definitely is not.

**Reply by ruanformigoni** [reply]
No problem! Thanks for the reply :smile:

**Reply by mhx** [reply]
FWIW, I've got my work branch building fine on Alpine. I might try switching to Alpile/musl for the static release binaries.

**Reply by mhx** [reply]
The new v0.12.0 release binaries are linked against `musl`. However, keep in mind that they're also linked against libfuse, which is also LGPL (I think), so there's no real win here from the licensing perspective.

**Comment by brakhane** [comment]
@ruanformigoni Sorry for chiming in, I just stumbled upon this discussion.
> Lastly, would you consider compiling with `musl libc` instead of `glibc`? The `glibc` binary works, but it has a licensing issue since it is provided as LGPL, so the object code must be provided and static linking voids that ([here](https://opensource.stackexchange.com/questions/13588/how-should-i-license-my-software-if-it-links-statically-with-glibc) is a summary). I have implemented some time ago a [musl dockerfile](https://github.com/ruanformigoni/dwarfs/blob/main/docker/Dockerfile.alpine) for dwarfs. My memory is a bit faint on that right now, but I could help to change the project libc to `musl libc` if you'd like.

As it currently stands, dwarfs is licensed under the GPLv3, which has even tighter restrictions, you can't include it in Apache licensed software.

@mhx What are your thoughts about releasing a library that can read dwarfs under say Apache or MIT license? I'm thinking about implementing such a library because I believe Dwarfs could make a good container format for things like game engines. Since the format is documented, I assume it's ok with you, but would like to make sure I'm not stepping on your toes.

Of course, if dwarfs itself would have the decoding part under a more liberal license, it would make stuff easier when it comes to keeping up to date with improvements to the format, but I think it could be difficult in practise.

FWIW, I believe keeping mkdwarfs and the whole creation part under GPL is a good idea, to keep (most) further improvements by other parties available to all.

**Reply by ruanformigoni** [reply]
@brakhane Thanks for the info, I'll have to think about changing the license on my side. From what I understood, distributing the binaries from the releases is allowed if the source code (or a clear method to obtain it) is also provided, maybe I'll just keep it that way for now.

**Reply by brakhane** [reply]
@ruanformigoni Yes, that is allowed. You can even sell GPL software. The restriction is that you need to give every recipient of the software the same rights you enjoyed: to analyse, modify, and distribute unchanged as well as derivative works, which must also be under the GPL.

So flatimage for example would need to be GPL if you include (parts of) dwarfs code in it, though the software that is packed with it wouldn't need to be, as long as flatimage is not required for the software to function. It could be Apache 2 if you implemented a "dwarfs reader" yourself that doesn't use the code from DwarFS at all. The output of GPL programs are not covered by the GPL.

**Reply by mhx** [reply]
> @mhx What are your thoughts about releasing a library that can read dwarfs under say Apache or MIT license?

Definitely worth thinking about! The decoding part is straightforward and most of the "magic" does indeed happen in the encoder. I'd have to dig into the code to see if I'm missing something obvious, but I'm certainly not opposed to the idea.

**Comment by mhx** [comment]
Please take a look at [dwarfs-0.10.0](https://github.com/mhx/dwarfs/releases/tag/v0.9.10) which I've just published. It's the first release that [properly exposes the libraries](https://github.com/mhx/dwarfs?tab=readme-ov-file#using-the-libraries). There's also some [example code](https://github.com/mhx/dwarfs/tree/main/example).

The libraries are split into `dwarfs_common`, `dwarfs_reader`, `dwarfs_writer`, `dwarfs_extractor` and `dwarfs_rewrite`. For reading a DwarFS image, you only need `dwarfs_reader` and `dwarfs_common` (which will be automatically pulled in as a dependency when using the CMake config).

Licensing `dwarfs_common` and `dwarfs_reader` more liberally in the future is certainly a possibility.

**Reply by brakhane** [reply]
Thanks, I'll check it out

**Comment by silentnoodlemaster** [comment]
Is it possible to configure cmake to not create the library files when building dwarfs?

**Reply by mhx** [reply]
Do you mean *create* or *install*? :)

The libraries *must* be created, as all tools link against them. *If* the libraries are built/linked statically, there's no need to actually install the libraries (but right now that's not an option).

I can consider adding an option for this in a future release. Until then, you could try replacing

```
if(NOT STATIC_BUILD_DO_NOT_USE)
```

in `cmake/libdwarfs.cmake` with

```
if(OFF)
```

and making sure you pass `-DBUILD_SHARED_LIBS=OFF` to `cmake`.

**Reply by silentnoodlemaster** [reply]
oh yes, i meant installed. it would be convenient to have it as an option

**Comment by silentnoodlemaster** [comment]
on the other hand, making a development package that contains only those files, could some flags be used to skip unneeded steps?

btw would these be the files?
```
   /usr/include/dwarfs/block_compressor.h
   /usr/include/dwarfs/block_compressor_parser.h
   /usr/include/dwarfs/checksum.h
   /usr/include/dwarfs/compiler.h
   /usr/include/dwarfs/compression.h
   /usr/include/dwarfs/compression_constraints.h
   /usr/include/dwarfs/config.h
   /usr/include/dwarfs/conv.h
   /usr/include/dwarfs/error.h
   /usr/include/dwarfs/file_access.h
   /usr/include/dwarfs/file_access_generic.h
   /usr/include/dwarfs/file_stat.h
   /usr/include/dwarfs/file_type.h
   /usr/include/dwarfs/file_util.h
   /usr/include/dwarfs/format.h
   /usr/include/dwarfs/fstypes.h
   /usr/include/dwarfs/history.h
   /usr/include/dwarfs/history_config.h
   /usr/include/dwarfs/integral_value_parser.h
   /usr/include/dwarfs/library_dependencies.h
   /usr/include/dwarfs/logger.h
   /usr/include/dwarfs/map_util.h
   /usr/include/dwarfs/match.h
   /usr/include/dwarfs/mmap.h
   /usr/include/dwarfs/mmif.h
   /usr/include/dwarfs/option_map.h
   /usr/include/dwarfs/os_access.h
   /usr/include/dwarfs/os_access_generic.h
   /usr/include/dwarfs/pcm_sample_transformer.h
   /usr/include/dwarfs/performance_monitor.h
   /usr/include/dwarfs/reader/block_cache_options.h
   /usr/include/dwarfs/reader/block_range.h
   /usr/include/dwarfs/reader/cache_tidy_config.h
   /usr/include/dwarfs/reader/filesystem_options.h
   /usr/include/dwarfs/reader/filesystem_v2.h
   /usr/include/dwarfs/reader/fsinfo_features.h
   /usr/include/dwarfs/reader/fsinfo_options.h
   /usr/include/dwarfs/reader/getattr_options.h
   /usr/include/dwarfs/reader/inode_reader_options.h
   /usr/include/dwarfs/reader/iovec_read_buf.h
   /usr/include/dwarfs/reader/metadata_options.h
   /usr/include/dwarfs/reader/metadata_types.h
   /usr/include/dwarfs/reader/mlock_mode.h
   /usr/include/dwarfs/scope_exit.h
   /usr/include/dwarfs/small_vector.h
   /usr/include/dwarfs/string.h
   /usr/include/dwarfs/terminal.h
   /usr/include/dwarfs/terminal_ansi.h
   /usr/include/dwarfs/thread_pool.h
   /usr/include/dwarfs/types.h
   /usr/include/dwarfs/util.h
   /usr/include/dwarfs/utility/filesystem_extractor.h
   /usr/include/dwarfs/utility/rewrite_filesystem.h
   /usr/include/dwarfs/utility/rewrite_options.h
   /usr/include/dwarfs/version.h
   /usr/include/dwarfs/vfs_stat.h
   /usr/include/dwarfs/writer/categorized_option.h
   /usr/include/dwarfs/writer/categorizer.h
   /usr/include/dwarfs/writer/category_parser.h
   /usr/include/dwarfs/writer/category_resolver.h
   /usr/include/dwarfs/writer/chmod_entry_transformer.h
   /usr/include/dwarfs/writer/compression_metadata_requirements.h
   /usr/include/dwarfs/writer/console_writer.h
   /usr/include/dwarfs/writer/contextual_option.h
   /usr/include/dwarfs/writer/entry_factory.h
   /usr/include/dwarfs/writer/entry_filter.h
   /usr/include/dwarfs/writer/entry_interface.h
   /usr/include/dwarfs/writer/entry_transformer.h
   /usr/include/dwarfs/writer/filesystem_block_category_resolver.h
   /usr/include/dwarfs/writer/filesystem_writer.h
   /usr/include/dwarfs/writer/filesystem_writer_options.h
   /usr/include/dwarfs/writer/filter_debug.h
   /usr/include/dwarfs/writer/fragment_category.h
   /usr/include/dwarfs/writer/fragment_order_options.h
   /usr/include/dwarfs/writer/fragment_order_parser.h
   /usr/include/dwarfs/writer/inode_fragments.h
   /usr/include/dwarfs/writer/inode_options.h
   /usr/include/dwarfs/writer/object.h
   /usr/include/dwarfs/writer/rule_based_entry_filter.h
   /usr/include/dwarfs/writer/scanner.h
   /usr/include/dwarfs/writer/scanner_options.h
   /usr/include/dwarfs/writer/segmenter.h
   /usr/include/dwarfs/writer/segmenter_factory.h
   /usr/include/dwarfs/writer/writer_progress.h
   /usr/include/dwarfs/xattr.h
   /usr/include/dwarfs/zstd_context_manager.h
   /usr/lib/debug/usr/lib/libdwarfs_common.so.0.10.1-0.10.1-1.fc40.x86_64.debug
   /usr/lib/debug/usr/lib/libdwarfs_extractor.so.0.10.1-0.10.1-1.fc40.x86_64.debug
   /usr/lib/debug/usr/lib/libdwarfs_reader.so.0.10.1-0.10.1-1.fc40.x86_64.debug
   /usr/lib/debug/usr/lib/libdwarfs_rewrite.so.0.10.1-0.10.1-1.fc40.x86_64.debug
   /usr/lib/debug/usr/lib/libdwarfs_writer.so.0.10.1-0.10.1-1.fc40.x86_64.debug
   /usr/lib/libdwarfs_common.so
   /usr/lib/libdwarfs_common.so.0.10.1
   /usr/lib/libdwarfs_extractor.so
   /usr/lib/libdwarfs_extractor.so.0.10.1
   /usr/lib/libdwarfs_reader.so
   /usr/lib/libdwarfs_reader.so.0.10.1
   /usr/lib/libdwarfs_rewrite.so
   /usr/lib/libdwarfs_rewrite.so.0.10.1
   /usr/lib/libdwarfs_writer.so
   /usr/lib/libdwarfs_writer.so.0.10.1
   /usr/lib64/cmake/dwarfs/dwarfs-config-version.cmake
   /usr/lib64/cmake/dwarfs/dwarfs-config.cmake
   /usr/lib64/cmake/dwarfs/dwarfs-targets-release.cmake
   /usr/lib64/cmake/dwarfs/dwarfs-targets.cmake
   ```

**Reply by mhx** [reply]
That looks about right. Actually, there's [support for building the libraries, the tools and the FUSE driver separately](https://github.com/mhx/dwarfs?tab=readme-ov-file#building). So you could first just build/install the libraries part, then the tools and/or FUSE driver.

As for packaging, I would assume there'd be a `libdwarfs` and a `libdwarfs-dev` with `libdwarfs` containing just `/usr/lib/libdwarfs_*.so*` and `libdwarfs-dev` everything else under `/usr/lib*` as well as `/usr/include`. Not sure how man pages are handled these days, but `dwarfs-format.5` should also be part of the libraries install.
